src/wrapper/clap-juce-wrapper.cpp

```diff
@@ -819,176 +824,219 @@ class ClapJuceWrapper : public clap::helpers::Plugin<
-        busses.fill(nullptr);
+        auto processEvent = [&](int sampleOffset) {
+            auto event = events->get(events, (uint32_t)currentEvent);
+            process_clap_event(event, sampleOffset);
 
-        /*DBG("IO Configuration: I=" << (int)process->audio_inputs_count << " O="
-                                   << (int)process->audio_outputs_count << " MX=" << (int)mx);
-        DBG("Plugin Configuration: IC=" << processor->getTotalNumInputChannels()
-                                        << " OC=" << processor->getTotalNumOutputChannels());
-        */
+            currentEvent++;
+            nextEventTime = (currentEvent < numEvents)
+                                ? (int)events->get(events, (uint32_t)currentEvent)->time
+                                : numSamples;
+        };
 
         /*
          * OK so here is what JUCE expects in its audio buffer. It *always* uses input as output
          * buffer so we need to create a buffer where each channel is the channel of the associated
          * output pointer (fine) and then the inputs need to either check they are the same or copy.
          */
+        static constexpr uint32_t maxBuses = 128;
+        std::array<float *, maxBuses> busses{};
+        busses.fill(nullptr);
 
-        /*
-         * So first lets load up with our outputs
-         */
-        uint32_t ochans = 0;
-        for (uint32_t idx = 0; idx < process->audio_outputs_count && ochans < maxBuses; ++idx)
+        for (int n = 0; n < numSamples;)
         {
-            for (uint32_t ch = 0; ch < process->audio_outputs[idx].channel_count; ++ch)
+            while (nextEventTime == n) // process all events at this timestep
+                processEvent(n);
+
+            auto getSamplesToProcess = [&]() {
+                if (CLAP_EVENT_RESOLUTION_SAMPLES <= 0)
+                {
+                    // Sample-accurate events are turned off, so just process the whole block!
+                    return numSamples;
+                }
+                else
+                {
+                    // How many samples should we process at a time?
+                    // In the spirit of sample-accurate events, we want to process a batch of
+                    // samples until we hit the next event, but we don't want to have a batch
+                    // smaller than the `CLAP_EVENT_RESOLUTION_SAMPLES`. If there's no more
+                    // events, just process the rest of the block!
+                    return (numSamples - n >= CLAP_EVENT_RESOLUTION_SAMPLES)
+                               ? juce::jmax(nextEventTime - n, CLAP_EVENT_RESOLUTION_SAMPLES)
+                               : (numSamples - n);
+                }
+            };
+
+            const auto numSamplesToProcess = getSamplesToProcess();
+
+            // process the rest of the events in this sub-block
+            while (nextEventTime < n + numSamplesToProcess && currentEvent < numEvents)
```

Comment:
This while seems like the more fulsome one vs the one above. At least add a comment why there are two?